compiler/optimizations/optimizeOnClauses.cpp

```diff
@@ -411,30 +446,77 @@ optimizeOnClauses(void) {
   if (fNoOptimizeOnClauses)
     return;
 
+  // If we're not using locks, the extern functions in
+  // atomics are local and safe for fast on.
+  if (0 != strcmp(CHPL_ATOMICS, "locks")) {
+    forv_Vec(ModuleSymbol, module, gModuleSymbols) {
+      if( module->hasFlag(FLAG_ATOMIC_MODULE) ) {
+        Vec<FnSymbol*> moduleFunctions = module->getTopLevelFunctions(true);
+        forv_Vec(FnSymbol, fn, moduleFunctions) {
+          if( fn->hasFlag(FLAG_EXTERN) ) {
+            fn->addFlag(FLAG_FAST_ON_SAFE_EXTERN);
+            fn->addFlag(FLAG_LOCAL_FN);
+          }
+        }
+      }
+    }
+  }
```

Comment:
As these intrinsics are not standardized, it can depend on the compiler.
For GCC at least, the documentation says:

    Not all operations are supported by all target processors. If a particular operation cannot be implemented on the target processor, a warning will be generated and a call an external function will be generated. The external function will carry the same name as the builtin, with an additional suffix `_n' where n is the size of the data type.

(from https://gcc.gnu.org/onlinedocs/gcc-4.4.5/gcc/Atomic-Builtins.html)

I'm pretty sure that any version of GCC with these intrinsics does not
silently re-implement them with a lock. We could check docs for other
versions of GCC or other compilers.

In fact that would not generally be reasonable for a compiler to
implement these intrinsics with a lock because the intrinsics work with a
pointer of the atomic object size (e.g. int*, int16_t*). If they were
implemented with a lock, the most reasonable implementation would use a
different data type. So, I think the only practical implementation a
compiler might have built would be a global lock... but that would
totally defeat the purpose of the intrinsics.